unblob/handlers/archive/arj.py

```diff
@@ -0,0 +1,184 @@
+            uint8 garble_password;
+            uint32 datetime_modified;
+            uint32 compressed_size;
+            uint32 original_size;
+            uint32 original_file_crc;
+            uint16 entryname_pos_in_filename;
+            uint16 file_access_mode;
+            uint16 host_data;
+        };
+
+        struct extended_header {
+            ushort size;
+        }
+    """
+
+    HEADER_STRUCT = "arj_header"
+
+    def _read_arj_main_header(
+        self, cparser: cstruct, file: io.BufferedIOBase, start_offset: int
+    ) -> int:
+        file.seek(start_offset)
+        basic_header = cparser.basic_header(file)
+
+        # It's unlikely and unhelpful if we find a completely zero-sized ARJ on it's own,
+        # so we raise here.
+        if basic_header.size == 0:
+            raise ARJNullFileException
+
+        file.seek(start_offset)
+        main_header = cparser.arj_header(file)
+        file.seek(start_offset + main_header.first_hdr_size + len(cparser.basic_header))
+
+        # Read main ARJ file name, comment, and CRC
+        _ = read_until_null(file)
+        _ = read_until_null(file)
+        _ = file.read(4)
+
+        # TODO: Handle the extended header
+        extended_header = cparser.extended_header(file)
+        if extended_header.size != 0:
+            raise ARJExtendedHeaderException
+
+        return file.tell()
+
+    def _read_arj_files(
+        self, cparser: cstruct, file: io.BufferedIOBase, start_of_files: int
+    ) -> int:
+        while True:
+            # Save where we're starting
+            start = file.tell()
+            basic_header = cparser.basic_header(file)
+
+            if basic_header.size == 0:
+                # We've reached the final empty file header. This is where we want to be.
+                return file.tell()
+
+            # Cursor back to the start of the files section.
+            file.seek(start)
+            file_header = cparser.file_header(file)
+
+            # Use the first_hdr_size to read past the header to the filename and comment strings.
+            file.seek(start + len(cparser.basic_header) + file_header.first_hdr_size)
+
+            # Read the filename, comment, and CRC
+            _ = read_until_null(file)
+            _ = read_until_null(file)
+            _ = file.read(4)
+
+            # TODO: Handle cases where there is an extended header.
+            extended_header = cparser.extended_header(file)
+            if extended_header.size != 0:
+                raise ARJExtendedHeaderException
+
+            # Read past the file contents
+            _ = file.read(file_header.compressed_size)
+
+    def calculate_chunk(
+        self, file: io.BufferedIOBase, start_offset: int
+    ) -> Union[ValidChunk, UnknownChunk]:
+
+        # Not using the inbuild self.parse_header() here, because we need access to a few different
+        # structures at various points.
+        cparser = cstruct()
+        cparser.load(self.C_STRUCTURES)
```

Comment:
We should discuss this with @kissgyorgy and see if we can make some generalization.